File: 3740-minimum-distance-between-three-equal-elements-i/3740-minimum-distance-between-three-equal-elements-i.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <climits>

using namespace std;

class Solution {
public:
    int minimumDistance(vector<int>& nums) {
        // Map each value to the list of its indices
        unordered_map<int, vector<int>> pos;
        for (int i = 0; i < nums.size(); ++i) {
            pos[nums[i]].push_back(i);
        }

        int min_diff = INT_MAX;

        // Check 3 consecutive occurrences for each distinct value
        for (const auto& [val, indices] : pos) {
            if (indices.size() >= 3) {
                for (size_t i = 0; i + 2 < indices.size(); ++i) {
                    min_diff = min(min_diff, indices[i + 2] - indices[i]);
                }
            }
        }

        return (min_diff == INT_MAX) ? -1 : 2 * min_diff;
    }
};

File: 3740-minimum-distance-between-three-equal-elements-i/3740-minimum-distance-between-three-equal-elements-i.py
from collections import defaultdict

class Solution:
    def minimumDistance(self, nums: list[int]) -> int:
        indices = defaultdict(list)
        for idx, val in enumerate(nums):
            indices[val].append(idx)
            
        min_dist = float('inf')
        
        for idx_list in indices.values():
            if len(idx_list) >= 3:
                # Check consecutive triplets (i, j, k)
                for t in range(len(idx_list) - 2):
                    dist = 2 * (idx_list[t + 2] - idx_list[t])
                    if dist < min_dist:
                        min_dist = dist
                        
        return min_dist if min_dist != float('inf') else -1